KEP-2837: PodLevelResources changes for 1.33 alpha stage (#5145)
kubernetes/enhancements: keps/sig-node/2837-pod-level-resource-spec/README.md

    - [Admission Controller](#admission-controller)
    - [Eviction Manager](#eviction-manager)
    - [Pod Overhead](#pod-overhead)
    - [Hugepages](#hugepages)
    - [Memory Manager](#memory-manager)
    - [In-Place Pod Resize](#in-place-pod-resize)
      - [API changes](#api-changes)
      - [Resize Restart Policy](#resize-restart-policy)
      - [Implementation Details](#implementation-details)

>     - [[Scopred for Beta] CPU Manager](#scopred-for-beta-cpu-manager)

Reviewer: nit: scoped*

    - [[Scoped for Beta] Topology Manager](#scoped-for-beta-topology-manager)
    - [[Scoped for Beta] User Experience Survey](#scoped-for-beta-user-experience-survey)
    - [[Scoped for Beta] Surfacing Pod Resource Requirements](#scoped-for-beta-surfacing-pod-resource-requirements)
      - [The Challenge of Determining Effective Pod Resource Requirements](#the-challenge-of-determining-effective-pod-resource-requirements)
      - [Goals of surfacing Pod Resource Requirements](#goals-of-surfacing-pod-resource-requirements)

>       - [Implementation Details](#implementation-details-1)

Reviewer 2: nit: why implementation-details-1?

Reviewer 3: Because there's another "implementation details" on line 41. This is just how github generates the anchors.

}
```

#### PodStatus API changes

Extend `PodStatus` to include pod-level analog of the container status resource
fields. Pod-level resource information in `PodStatus` is essential for pod-level [In-Place Pod
Update]
(https://github.com/kubernetes/enhancements/blob/master/keps/sig-node/1287-in-place-update-pod-resources/README.md#api-changes)
as it provides a way to track, report and use the actual resource allocation for the
pod, both before and after a resize operation.

```
type PodStatus struct {
...
  // Resources represents the compute resource requests and limits that have been

>   // applied at the pod level. If pod-level resources are not explicitly specified,

Reviewer 4: Are only the resources supported in pod-level spec.resources (cpu, memory, and now hugepages...) aggregated here, or are other custom resources specified in the containers aggregated here? (as an aside, I think the pod-level resource validation errors if container-level resources are specified which are not included in pod-level resources at all... https://github.com/kubernetes/kubernetes/blob/ee22760391bae28954a69dff499d1cead9a9fcf0/pkg/apis/core/validation/validation.go#L4340-L4356).

What happens if pod-level spec.resources sets a pod-level cpu limit, but not a memory limit, and individual containers all set a memory limits. Does this include the pod-level cpu limit and the aggregated container memory limits?

Reviewer 2: For those extended resources, this is still an open question. Luckily we can address this in later releases. Not a blocking for 1.33.

Reviewer 2: cc/ @yujuhong @johnbelamaric

Reviewer 5: > For those extended resources, this is still an open question. Luckily we can address this in later releases.

Ack. I think this should be stated in the non-goals if they are not already there.

Author: It is stated in the non-goals section that only CPU, memory and hugepages are supported for now https://github.com/kubernetes/enhancements/blob/master/keps/sig-node/2837-pod-level-resource-spec/README.md#non-goals

Also, what Jordan pointed out there's a bug in validation logic where if container-level resources are set for unsupported resource type, the validation logic will error out because aggregated container requests in this case will be greater than pod requests ( as pod-level resources won't be set for unsupported resources): https://github.com/kubernetes/kubernetes/blob/ee22760391bae28954a69dff499d1cead9a9fcf0/pkg/apis/core/validation/validation.go#L4340-L4356

I will fix the bug. Thanks @liggitt for finding the bug.

Author: https://github.com/kubernetes/kubernetes/pull/130131 is the PR @liggitt

  // then these will be the aggregate resources computed from containers. If limits are

>   // not defined for all containers (and not defined at the pod level) then no aggregate

Reviewer 4: is this saying
1. limits will not be *populated* in this field if they're not specified in spec.resources and we can't aggregate them from containers because not all containers specify limits
2. limits *will* be populated here but not applied/enforced at the pod level?

Author: Updated the language to clarity

>   // not defined for all containers (and not defined at the pod level) then no aggregate

Author: the pod level limits serve as a hard cap.

  // pod-level limits will be applied.
  // +featureGate=InPlacePodVerticalScaling
  // +optional

>   Resources *ResourceRequirements

Author: We used the same type i.e. ResourceRequirements for Resources in PodSpec as well.

>   Resources *ResourceRequirements

Author: Do we want to add a new type ResourceConstraints or ResourceRequestsLimits?

  // AllocatedResources is the total requests allocated for this pod by the node.
  // Kubelet sets this to the accepted requests when a pod (or resize) is admitted.
  // If pod-level requests are not set, this will be the total requests aggregated
  // across containers in the pod.
  // +featureGate=InPlacePodVerticalScaling
  // +optional

>   AllocatedResources ResourceList

Reviewer 4: pointer for new optional field?

```suggestion
  AllocatedResources *ResourceList
```

Reviewer 4: Ah, I didn't chase the type. As long as it doesn't serialize anything in JSON or proto when empty, that's fine

}
```

#### PodStatus API changes

> Extend `PodStatus` to include pod-level analog of the container status resource

Reviewer 3: It should, but the intermediate "requested" is just Kubelet internal implementation, not surfaced in the API.

  size of the pod's cgroup. This means the pod cgroup's resource limits will be
  set to accommodate both pod-level requests and pod overhead.

#### Hugepages

> With the proposed changes, hugepages-2Mi and hugepages-1Gi will be added to the pod-level resources section, alongside CPU and memory. The hugetlb cgroup for the

Reviewer 4: only these two hugepages sizes specifically, or anything matching `hugepages-` prefix pattern that passes hugepages name and value validation (just like container limits/requests)?

https://github.com/kubernetes/kubernetes/blob/ee22760391bae28954a69dff499d1cead9a9fcf0/pkg/apis/core/validation/validation.go#L7099

> pod will then directly reflect the pod-level hugepage limits, rather than using an aggregated value from container limits. When scheduling, the scheduler will 

Reviewer 4: I'm not that familiar with today's behavior... are individual containers not constrained to the hugepages limits they specify, but can steal allocated hugepages resources from a shared space sized to fit the other containers in the pod's hugepages requests?

Author: With the current behavior, the containers are constrained are constrained to hugepages limits they specify. With pod-level resources, we want to enable the containers to use huge pages from a shared huge pages pool in a pod.

Manager will evolve to track and enforce resource limits at both the pod and
container levels. It will need to aggregate memory usage across all containers
within a pod to calculate the pod's total memory consumption. The Memory Manager
will then enforce the pod-level limit as the hard cap for the entire pod's memory
usage, preventing it from exceeding the allocated amount.  While still
maintaining container-level limit enforcement, the Memory Manager will need to
coordinate with the Kubelet and eviction manager to make decisions about pod
eviction or individual container termination when the pod-level limit is
breached.

#### CPU Manager

With the introduction of pod-level resource specifications, the CPU manager in
Kubernetes will adapt to manage CPU requests and limits at the pod level rather
than solely at the container level. This change means that the CPU manager will
allocate and enforce CPU resources based on the total requirements of the entire
pod, allowing for more flexible and efficient CPU utilization across all
containers within a pod. The CPU manager will need to ensure that the aggregate

> CPU usage of all containers in a pod does not exceed the pod-level limits. 

Reviewer 6: what about the exclusive allocation of CPUs? Is that something we keep at container level?

Author: Thanks for the questions @swatisehgal. Scoping out CPU manager also for Beta then as it requires thinking about all the different policy options before finalizing the design. The main focus of alpha 2 in 1.33 is in-place pod related API changes so removing other complex features out of scope of alpha 2 then. I will revisit these for Beta after discussions offline and in sig node meetings.

Reviewer 6: I'm late to the party, but @swatisehgal captured the most pressing points anyway. Looking forward to join the future conversations in sig-nodel.

  resize of the pod-level memory limit doesn't change the effective container limit,
  so the c2 is not restarted when the pod-level memory limit is resized.

##### Implementation Details

###### Allocating Pod-level Resources
Allocation of pod-level resources will work the same as container-level resources. The allocated resources checkpoint will be extended to include pod-level resources, and the pod object will be updated with the allocated resources in the pod sync loop.

###### Actuating Pod-level Resource Resize
A dirty bit for pod-level resources will be added to the allocation checkpoint to
signal the need for resize.  This change, along with the existing tracking and
triggering mechanisms, is sufficient for pod-level resize actuation.  The same
ordering rules for pod and container resource resizing will be applied for each
resource as needed:
1. Increase pod-level cgroup (if needed)
2. Decrease container resources
3. Decrease pod-level cgroup (if needed)
4. Increase container resources

###### Tracking Actual Pod-level Resources
To accurately track actual pod-level resources during in-place pod resizing, several

> changes are required:

Reviewer 4: are these the ~same as the container-level resizes in terms of what gets "accepted" by the kubelet and reported back in API status before and after the resize actually happens?

Author: Yes, all the changes in this KEP to support pod-level resources resizing is same as changes made for container-level resources resizing.

1. Configuration reading: Pod-level resource config is currently read as part of the
   resize flow, but will also need to be read during pod creation. Critically, the
   configuration must be read again after the resize operation to capture the
   updated resource values.  Currently, the configuration is only read before a

>    resize.

Author: I meant to say, in alpha, always re-read pod-level resources from the API server in the sync loop. And then evaluate caching in Beta. Does that make sense?

>    resize.

Author: updated, PTAL

> 2. Pod Status Update: Because the pod status is updated before the resize takes

Reviewer 3: Aside: We should probably re-evaluate this, outside the context of this KEP. Now that there are things reflected in the status that don't also trigger resync, we're going to need to resync the pod just to write another field to the status. I'm not sure off hand what the consequences of moving the status update to the end of PodSync would be.

   effect, the status will not immediately reflect the new resource values.  If a
   container within the pod is also being resized, the container resize operation
   will trigger a pod synchronization (pod-sync), which will refresh the pod's
   status.  However, if only pod-level resources are being resized, a pod-sync must
   be explicitly triggered to update the pod status with the new resource
   allocation.

> 3. [Scoped for Beta] Caching: Actual pod resource data may be cached to minimize API server load. This cache, if implemented, must be invalidated after each successful pod resize to ensure that subsequent reads retrieve the latest information. The need for and implementation of this caching mechanism will be evaluated in the beta phase.  Performance benchmarking will be conducted to determine if caching is required and, if so, what caching strategy is most appropriate.

Reviewer 3: I don't understand this item, it seems at odds with [1]? If we're not caching the configuration, then it needs to be re-read with each sync loop, so [1] is irrelevant?

I think the question is, should we cache in-memory (and refresh the cache based on the conditions from [1]), or should we just re-read the values on every pod-sync (if it's sufficiently fast to do so)?

Author: Point 3 says caching is scoped for beta

Author: updated the text. PTAL

Reviewer 7: Reads by whom?

Author: By Kubelet. Updated the text.

**Note for future enhancements for Ephemeral containers with pod-level resources and
IPPR**
Previously, assigning resources to ephemeral
containers wasn't allowed because pod resource allocations were immutable. With

> the introduction of in-place pod resizing, users will gain more flexibility:

Reviewer 3: nit: since we're talking about a future enhancement, not the current proposal
```suggestion
the introduction of in-place pod resizing, users could gain more flexibility:
```

* Adjust pod-level resources to accommodate the needs of ephemeral containers. This
allows for a more dynamic allocation of resources within the pod.
* Specify resource requests and limits directly for ephemeral containers. Kubernetes will
then automatically resize the pod to ensure sufficient resources are available
for both regular and ephemeral containers.

Currently, setting `resources` for ephemeral containers is disallowed as pod
resource allocations were immutable before In-Place Pod Resizing feature. With
in-place pod resize for pod-level resource allocation, users should be able to
either modify the pod-level resources to accommodate ephemeral containers or
supply resources at container-level for ephemeral containers and kubernetes will
resize the pod to accommodate the ephemeral containers.

#### [Scoped for Beta] Topology Manager

Note: This section includes only high level overview; Design details will be added in Beta stage.

* The pod level scope for topology aligntment will consider pod level requests and limits instead of container level aggregates.
* The hint providers will consider pod level requests and limits instead of

>   container level aggregates.

Reviewer 6: How do we handle the current use case though? if the user does _not_ set pod-level resources, will we have 2 code paths (the existing + the new one being proposed), or do we want to automatically compute the pod-level resources aggregating the container-level resources, improving on what the code currently does?

For topology manager we will probably need some more details about the design

Author: - When pod-level resources are not specified, as in the current use case, we change nothing in the pod spec. Today, cgroups at pod-level are still calculated using the aggregation method. We will continue to have that path for pod spec without pod level resources. So to summarize, pod-level resources are calculated even today and used in cgroup values at pod-level, but not updated in pod-spec. 

- We definitely need more details around design for topology manager which is why I intentionally want to skip it for alpha and focus on CPU manager and memory manager first for alpha.

Reviewer 8: What about the case where pod level requests/limits are specified but `container` scope configured on the node? Do we disregard pod level requests/limits in that case?

Reviewer 8: Additionally, It might be obvious, but it's important to emphasize that in case pod level resources are specified we must consider the topology manager's policies and policy options configured on the node.

  size of the pod's cgroup. This means the pod cgroup's resource limits will be
  set to accommodate both pod-level requests and pod overhead.

#### Hugepages

> With the proposed changes, support for hugepages(with prefix hugepages-*) will be extended to the pod-level resources specifications, alongside CPU and memory. The hugetlb cgroup for the

Reviewer 9: ```suggestion
With the proposed changes, support for Linux hugepages (resources with prefix `hugepages-`) will be extended to the pod-level resources specifications, alongside CPU and memory. The hugetlb cgroup for the
```
?

pod will then directly reflect the pod-level hugepage limits, if specified, rather than using an aggregated value from container limits. When scheduling, the scheduler will 
consider hugepage requests at the pod level to find nodes with enough available
resources.

> Containers will still need to mount an emptyDir volume to access the huge page filesystem (typically /dev/hugepages).  This is the standard way for containers to interact with huge pages, and this will not change. 

Reviewer 9: This sounds confusing; `/dev/hugepages` isn't normally present within an empty directory (per `emptyDir`). Maybe clarify what we mean here.

#### Memory Manager

With the introduction of pod-level resource specifications, the Kubernetes Memory
Manager will evolve to track and enforce resource limits at both the pod and
container levels. It will need to aggregate memory usage across all containers
within a pod to calculate the pod's total memory consumption. The Memory Manager
will then enforce the pod-level limit as the hard cap for the entire pod's memory
usage, preventing it from exceeding the allocated amount.  While still
maintaining container-level limit enforcement, the Memory Manager will need to
coordinate with the Kubelet and eviction manager to make decisions about pod
eviction or individual container termination when the pod-level limit is
breached.

#### In-Place Pod Resize

##### API changes

IPPR for pod-level resources requires extending `PodStatus` to include pod-level

> resource fields as detailed in [PodStatus API changes](#### PodStatus API changes)

Reviewer 9: TODO

Reviewer 3: ```suggestion
resource fields as detailed in [PodStatus API changes](#podstatus-api-changes)
```

section.

##### Resize Restart Policy

Pod-level resize policy is not supported in the alpha stage of Pod-level resource
feature. While a pod-level resize policy might be beneficial for VM-based runtimes
like Kata Containers (potentially allowing the hypervisor to restart the entire VM
on resize), this is a topic for future consideration. We plan to engage with the
Kata community to discuss this further and will re-evaluate the need for a pod-level
policy in subsequent development stages.

The absence of a pod-level resize policy means that container restarts are
exclusively managed by their individual `resizePolicy` configs. The example below of
a pod with pod-level resources demonstrates several key aspects of this behavior,
showing how containers without explicit limits (which inherit pod-level limits) interact
with resize policy, and how containers with specified resources remain unaffected by
pod-level resizes.

```yaml
apiVersion: v1
kind: Pod
metadata:
  name:     pod-level-resources
spec:
  resources:
    requests:
      cpu: 100m
      memory: 100Mi
    limits:
      cpu: 200m
      memory: 200Mi
  containers:
    - name: c1

>       image: registry.k8s.io/pause:latest

Reviewer 9: nit: can we avoid implying that `latest` is a good choice of version pin?

  resize of the pod-level memory limit doesn't change the effective container limit,
  so the c2 is not restarted when the pod-level memory limit is resized.

##### Implementation Details

###### Allocating Pod-level Resources
Allocation of pod-level resources will work the same as container-level resources. The allocated resources checkpoint will be extended to include pod-level resources, and the pod object will be updated with the allocated resources in the pod sync loop.

###### Actuating Pod-level Resource Resize
The mechanism for actuating pod-level resize remains largely unchanged from the
existing container-level resize process.  When pod-level resource configurations are
applied, the system handles the resize in a similar manner as it does for
container-level resources. This includes extending the existing logic to incorporate
directly configured pod-level resource settings.

The same ordering rules for pod and container resource resizing will be applied for each
resource as needed:
1. Increase pod-level cgroup (if needed)
2. Decrease container resources
3. Decrease pod-level cgroup (if needed)
4. Increase container resources

###### Tracking Actual Pod-level Resources

> To accurately track actual pod-level resources during in-place pod resizing, several

Reviewer 7: @tallclair Given the discussion on how NRI plugins or systemd can mutate the resources (e.g. rounding), what happens when:

* User specifies a pod with 2 containers, each using `5m` cpu and the pod is `10m`
* An NRI plugin mutates the containers and rounds them up to `10m` each

Are we smart enough to increase the pod?

Reviewer 3: Hmm, we don't today, but we could. We are reading the actual values from the runtime, so we could compute the the pod-level cgroups based on the sum of those instead of the allocated resources (or whichever is larger)

Reviewer 5: Hmm... but if NRI plugins change the value to be completely different, that'd just conflicts with how kubelet manages the cgroups. We can simply grab the values from the and assume those are the ones we want. 

@samuelkarp

Reviewer 2: Agreed with Yuju above. 

I am expecting that with those new efforts on the resource management we are doing and plan to do, the users eventually limit their NRI usages largely.

Reviewer 2: We can continue discussing this, but this isn't a blocker here.

   allocation.

3. [Scoped for Beta] Caching: Actual pod resource data may be cached to minimize API server load. This cache, if implemented, must be invalidated after each successful pod resize to ensure that subsequent reads retrieve the latest information. The need for and implementation of this caching mechanism will be evaluated in the beta phase.  Performance benchmarking will be conducted to determine if caching is required and, if so, what caching strategy is most appropriate.

**Note for future enhancements for Ephemeral containers with pod-level resources and
IPPR**
Previously, assigning resources to ephemeral
containers wasn't allowed because pod resource allocations were immutable. With
the introduction of in-place pod resizing, users could gain more flexibility:

* Adjust pod-level resources to accommodate the needs of ephemeral containers. This
allows for a more dynamic allocation of resources within the pod.
* Specify resource requests and limits directly for ephemeral containers. Kubernetes will
then automatically resize the pod to ensure sufficient resources are available
for both regular and ephemeral containers.

> Currently, setting `resources` for ephemeral containers is disallowed as pod

Reviewer 9: ```suggestion
Prior to this KEP, setting `resources` for ephemeral containers was disallowed as pod
```
(to clarify)

resource allocations were immutable before In-Place Pod Resizing feature. With
in-place pod resize for pod-level resource allocation, users should be able to
either modify the pod-level resources to accommodate ephemeral containers or
supply resources at container-level for ephemeral containers and kubernetes will
resize the pod to accommodate the ephemeral containers.

> #### [Scopred for Beta] CPU Manager

Reviewer 9: ```suggestion
#### [Scoped for Beta] CPU Manager
```

With the introduction of pod-level resource specifications, the CPU manager in
Kubernetes will adapt to manage CPU requests and limits at the pod level rather
than solely at the container level. This change means that the CPU manager will
allocate and enforce CPU resources based on the total requirements of the entire
pod, allowing for more flexible and efficient CPU utilization across all
containers within a pod. The CPU manager will need to ensure that the aggregate
CPU usage of all containers in a pod does not exceed the pod-level limits. 

The CPU Manager policies are container-level configurations that control the 
fine-grained allocation of CPU resources to containers.  While CPU manager 
policies will operate within the constraints of pod-level resource limits, they 
do not directly apply at the pod level.

#### [Scoped for Beta] Topology Manager

Note: This section includes only high level overview; Design details will be added in Beta stage.

> * The pod level scope for topology aligntment will consider pod level requests and limits instead of container level aggregates.

Reviewer 9: ```suggestion
* The pod level scope for topology alignment will consider pod level requests and limits instead of container level aggregates.
```
